pytorch_lightning/trainer/trainer.py

```diff
@@ -1334,3 +1335,11 @@ def _on_exception(self):
         # save a checkpoint for fault tolerant training. we don't use `log_dir` to minimize the chances of failure.
         file_path = os.path.join(self.default_root_dir, ".pl_auto_save.ckpt")
         self.save_checkpoint(file_path)
+
+    @contextmanager
+    def evaluation_context(self) -> Generator:
```

Comment:
would we also be happy with making this a utility function?
And I currently see no reason to include the check of `trainer.state.fn`.

Reply:
@ananthsub Do you think we should move this utility function somewhere else for reusability? In that case, should we leave the check against `fit` and `validation` loops here?

Reply:
@awaelchli @carmocca if we want to enable inference mode for validation in addition to testing and prediction, this becomes a super simple utility function.

I haven't found a good example for why this wouldn't work for validation, given we have no_grad already set.

One thing we need to allow for is users disabling this mode for any reason. To that end, we should document that this mode will be turned on by default for validation/testing/prediction.

Reply:
From the docs:

> Enable inference mode when you are performing computations that don’t need to be recorded in the backward graph, AND you don’t plan on using the tensors created in inference mode in any computation that is to be recorded by autograd later.

This constraint is not required by `no_grad`.

I also can't come up with a good example that breaks this constraint during fitting validation, but some users might need to

Reply:
```suggestion
    def _evaluation_context(self) -> Generator:
```

prefer keeping this private in the meantime.

For now I think this is the safest route. If we also want to enable this for validation, we easily can. Are there docs we should update in case users see any side effects with inference mode and need this disabled?

Reply:
@carmocca @ananthsub What do you think about this? Is there a better way to let users override the inference mode?

```python
@contextmanager
def _evaluation_context(self) -> Generator:
    with torch.inference_mode() if _TORCH_GREATER_EQUAL_1_9 and os.getenv(
        "NO_INFERENCE_MODE", "False"
    ).lower() == "false" and self.state.fn in (
        TrainerFn.TESTING,
        TrainerFn.PREDICTING,
    ) else torch.no_grad():
        yield
```

Reply:
@tangbinh I don't think we need a flag. we don't offer a similar one for no_grad. users are free to specify
```
with torch.inference_mode(mode=False):
    ....
```
inside of their code that needs to disable this